### DISCUSSION TITLE
Guider <=> PlateSolver dialog via shared memory and socket

### DISCUSSION THREAD

**Discussion by blumzi** [blumzi/MAST_unit#4] [Ideas]
We are writing a guider software that will acquire images and ask the PlateSolver to solve them

The current solver:
* Gets an image file and parameters on the command-line
* Loads catalogs
* Solves the image (may fail)
* Dies

We'd like to have a new scheme where the solver is a long running process spawned by our guider.  The protocol will be:
* The `guider` (started by our `unit.start_guiding` API call)
   * Opens a socket and listens to a well known port
   * Spawns the `solver` process
   * Waits for the `solver` to send a **ready** message
 * The `solver`, as soon as it is capable and willing (after loading catalogs), sends a **ready** message on the socket
 * Both sides loop until the guider send a **quit** message on the socket
    * The `guider` acquires an image (potentially after focusing)
       * Copies the image to a well known shared memory segment
       * Sends a **solve** message on the socket, including initial coordinates (from mount) and image width and height
       * Waits for a **solve-done** from `guider`
     * The `solver`
        * Receives the **solve** message, 
        * Solves the image
        * Sends a **solve-done** message including:
           * Success/failure
           * Solved coordinates
           * Reason for failure
           * Any additional relevant information
      *  The `guider`:
          * Saves the image to disk
          * If a mount correction is needed:
            * Send correction to the mount
            * Logs the correction for further reference
  * On the `unit.stop_guiding` API call:
     * The `guider` sends a **quit** message
     * The `solver` exits

**Comment by blumzi** [comment]
Arie has created a [github repository](https://github.com/blumzi/MAST-guider-for-kevin/tree/master) with an initial implementation of the `guider` and a PlateSolverSimulator.

We can use this repository to exchange ideas and implementations

**Comment by EranOfek** [comment]
The astrometric solution should appear as a standard WCS

To expedite the solution the user should control the astrometric search
radius.
In this case, in the first image of a field the search radius will be
large, while in all other solutions, it should be very small (~5"),
and revert to larger search only if failed.

**Comment by sagiph-weizmann** [comment]
Check with Oren what is the actual output.
